presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -349,6 +359,10 @@ private synchronized QueryResults getNextResult(OptionalLong token, UriInfo uriI
             }
         }
 
+        if (failed) {
+            checkState(queryInfo.getState().isDone() && queryInfo.getState() == QueryState.FAILED, "query should have failed");
```

Comment:
This looks super sketchy to me. I remember this code path being as racy as a P90D, and the errors would get super weird if query info is stale.  Are you sure that there is a happens before relationship between the query info update and the failure?

Reply:
You are right, checking `queryInfo.getState() == QueryState.FAILED` may be racy. But do you think we need to assert `queryInfo.getState().isDone()`? Two reasons: (1) `queryManager.getQueryInfo(queryId)` (15 lines above) will throw exception if `queryInfo` is null; so it is safe to call `getState()` and (2) `boolean failed = queryState.setIf(FAILED, currentState -> !currentState.isDone())` in `QueryStateMachine` guarantees the query should be in `Done` state after we call `failQuery`.

Of course, asserting `isDone()` might be redundant. I'm ok leaving or dropping it.